File: Sources/Kernel/VirtIOBlock.swift
// VirtIO Block Driver for SwiftOS

import CSupport

let VIRTIO_PCI_DEVICE_BLOCK: UInt16 = 0x1001

let VIRTIO_BLK_T_IN: UInt32 = 0
let VIRTIO_BLK_T_OUT: UInt32 = 1

struct VirtioBlkOuthdr {
    var type: UInt32
    var reserved: UInt32
    var sector: UInt64
}

struct VirtioBlkStatus {
    var status: UInt8
}

struct VirtioBlkDevice {
    var config: VirtioConfig
    var queueSize: Int = 0
    var desc: UnsafeMutablePointer<VirtqDesc>?
    var avail: UnsafeMutablePointer<VirtqAvail>?
    var used: UnsafeMutablePointer<VirtqUsed>?
    var lastUsedIdx: UInt16 = 0
}

nonisolated(unsafe) var blockDevice: VirtioBlkDevice?

func initVirtioBlock() {
    kprint("Block Probe\n")

    guard let dev = scanPci(vendor: VIRTIO_PCI_VENDOR, device: VIRTIO_PCI_DEVICE_BLOCK) else {
        kprint("Block NOT FOUND\n")
        return
    }

    kprint("Block Found\n")

    // Enable PCI Command register: IO Space (0), Memory Space (1), Bus Master (2)
    let cmd = pciRead16(bus: dev.0, slot: dev.1, funcNum: dev.2, offset: 4)
    pciWrite16(bus: dev.0, slot: dev.1, funcNum: dev.2, offset: 4, value: cmd | 0x7)

    var config = VirtioConfig()
    parseVirtioCapabilities(dev: dev, config: &config)

    if let common = config.common {
        kprint("Block Init\n")
        // 1. Reset
        common.pointee.device_status = 0
        // 2. Acknowledge
        common.pointee.device_status |= 1
        // 3. Driver
        common.pointee.device_status |= 2

        // 4. Features (negotiate)
        // For now just accept what the device offers
        common.pointee.device_status |= 8  // FEATURES_OK

        // 5. Setup Virtqueue
        common.pointee.queue_select = 0
        let qSize = Int(common.pointee.queue_size)
        kprint("  Queue Size: ")
        kprint_hex(UInt64(qSize))
        kprint("\n")

        // Allocate space for the virtqueue
        let vq_size = (qSize * 16 + 8 + qSize * 2 + 4095) & ~4095 + (8 + qSize * 8 + 4095) & ~4095
        let rawPtr = kernelAlloc(size: vq_size, align: 4096)

        let descPtr = rawPtr.assumingMemoryBound(to: VirtqDesc.self)
        let availPtr = rawPtr.advanced(by: qSize * 16).assumingMemoryBound(to: VirtqAvail.self)
        let usedPtr = rawPtr.advanced(by: (qSize * 16 + 8 + qSize * 2 + 4095) & ~4095)
            .assumingMemoryBound(to: VirtqUsed.self)

        common.pointee.queue_desc = UInt64(UInt(bitPattern: descPtr))
        common.pointee.queue_driver = UInt64(UInt(bitPattern: availPtr))
        common.pointee.queue_device = UInt64(UInt(bitPattern: usedPtr))
        common.pointee.queue_enable = 1

        // 6. Driver OK
        common.pointee.device_status |= 128
        kprint("Block READY\n")

        blockDevice = VirtioBlkDevice(
            config: config,
            queueSize: qSize,
            desc: descPtr,
            avail: availPtr,
            used: usedPtr
        )
    }
}

func virtioBlockRead(sector: UInt64, count: Int, buffer: UnsafeMutableRawPointer) -> Bool {
    if count <= 0 { return true }  // Nothing to do
    guard let dev = blockDevice, let descs = dev.desc, let avail = dev.avail else { return false }

    // Allocate space for request headers at a safe location
    let hdr = kernelAlloc(size: MemoryLayout<VirtioBlkOuthdr>.size).assumingMemoryBound(
        to: VirtioBlkOuthdr.self)
    let status = kernelAlloc(size: 1).assumingMemoryBound(to: UInt8.self)

    hdr.pointee.type = VIRTIO_BLK_T_IN
    hdr.pointee.reserved = 0
    hdr.pointee.sector = sector
    status.pointee = 0xFF

    descs[0].addr = UInt64(UInt(bitPattern: hdr))
    descs[0].len = UInt32(MemoryLayout<VirtioBlkOuthdr>.size)
    descs[0].flags = VIRTQ_DESC_F_NEXT
    descs[0].next = 1

    descs[1].addr = UInt64(UInt(bitPattern: buffer))
    descs[1].len = UInt32(count * 512)
    descs[1].flags = VIRTQ_DESC_F_NEXT | VIRTQ_DESC_F_WRITE
    descs[1].next = 2

    descs[2].addr = UInt64(UInt(bitPattern: status))
    descs[2].len = 1
    descs[2].flags = VIRTQ_DESC_F_WRITE
    descs[2].next = 0

    // Update avail ring
    let availIdx = Int(avail.pointee.idx % UInt16(dev.queueSize))
    let ringPtr = UnsafeMutablePointer<UInt16>(bitPattern: UInt(bitPattern: avail) + 4)!
    ringPtr[availIdx] = 0  // Head of descriptor chain

    // Memory barrier
    asm_volatile_barrier()

    avail.pointee.idx = avail.pointee.idx &+ 1

    // Notify
    if let notify = dev.config.notify, let common = dev.config.common {
        common.pointee.queue_select = 0
        let off = UInt(common.pointee.queue_notify_off) * UInt(dev.config.notify_off_multiplier)
        let notifyPtr = (notify + Int(off)).assumingMemoryBound(to: UInt16.self)
        notifyPtr.pointee = 0  // Queue index
    }

    // Wait for completion (poll used ring)
    var timeout = 10_000_000
    while dev.used!.pointee.idx == blockDevice!.lastUsedIdx && timeout > 0 {
        asm_pause()
        timeout -= 1
    }

    if timeout == 0 {
        kprint("Block Read TIMEOUT\n")
        return false
    }

    blockDevice?.lastUsedIdx = dev.used!.pointee.idx

    return status.pointee == 0
}

File: zig-kernel/src/virtio_block.zig
const virtio = @import("virtio.zig");

extern fn asm_pause() void;
extern fn asm_volatile_barrier() void;
extern fn kernel_alloc(size: usize, alignment: usize) ?*u8;

const VIRTIO_PCI_DEVICE_BLOCK: u16 = 0x1001;

const VIRTIO_BLK_T_IN: u32 = 0;
const VIRTIO_BLK_T_OUT: u32 = 1;

const VirtioBlkOuthdr = extern struct {
    typ: u32,
    reserved: u32,
    sector: u64,
};

const VirtioBlkStatus = extern struct { status: u8 };

const VirtioBlkDevice = struct {
    config: virtio.VirtioConfig,
    queueSize: usize,
    desc: ?*virtio.VirtqDesc,
    avail: ?*virtio.VirtqAvail,
    used: ?*virtio.VirtqUsed,
    lastUsedIdx: u16,
};

pub var blockDevice: ?*VirtioBlkDevice = null;

pub fn initVirtioBlock() void {
    // Probe
    const dev = virtio.scanPci(virtio.VIRTIO_PCI_VENDOR, VIRTIO_PCI_DEVICE_BLOCK) orelse return;

    // Enable PCI Command register: IO, Mem, Bus Master
    const cmd = virtio.pciRead16(dev.bus, dev.slot, dev.func, 4);
    virtio.pciWrite16(dev.bus, dev.slot, dev.func, 4, cmd | 0x7);

    var config: virtio.VirtioConfig = .{ .common = null, .notify = null, .isr = null, .device = null, .notify_off_multiplier = 0 };
    virtio.parseVirtioCapabilities(dev, &config);

    if (config.common) |common| {
        // Reset and driver handshake
        common.*.device_status = 0;
        common.*.device_status |= 1; // ACK
        common.*.device_status |= 2; // DRIVER
        common.*.device_status |= 8; // FEATURES_OK

        common.*.queue_select = 0;
        const qSize: usize = @intCast(common.*.queue_size);

        // Allocate virtqueue backing memory (desc + avail + used) with alignment
        const desc_bytes = qSize * 16;
        // const avail_bytes = 8 + qSize * 2;
        const used_bytes = 8 + qSize * 8;
        const vq_size = ((desc_bytes + 8 + qSize * 2 + 4095) & ~4095) + ((used_bytes + 4095) & ~4095);

        const rawPtr = kernel_alloc(vq_size, 4096) orelse return;

        const descPtr: *virtio.VirtqDesc = @ptrCast(rawPtr);
        const availPtr: *virtio.VirtqAvail = @ptrCast(@as(*virtio.VirtqAvail, @ptrFromInt(@intFromPtr(rawPtr) + desc_bytes)));
        const usedPtr: *virtio.VirtqUsed = @ptrCast(@as(*virtio.VirtqUsed, @ptrFromInt(@intFromPtr(rawPtr) + ((desc_bytes + 8 + qSize * 2 + 4095) & ~4095))));

        common.*.queue_desc = @as(u64, @intFromPtr(descPtr));
        common.*.queue_driver = @as(u64, @intFromPtr(availPtr));
        common.*.queue_device = @as(u64, @intFromPtr(usedPtr));
        common.*.queue_enable = 1;

        common.*.device_status |= 128; // DRIVER_OK

        // Allocate device struct storage
        const dev_ptr = kernel_alloc(@sizeOf(VirtioBlkDevice), 16) orelse return;
        const dev_struct: *VirtioBlkDevice = @ptrCast(dev_ptr);
        dev_struct.* = VirtioBlkDevice{ .config = config, .queueSize = qSize, .desc = descPtr, .avail = availPtr, .used = usedPtr, .lastUsedIdx = 0 };
        blockDevice = dev_struct;
    }
}

pub fn virtioBlockRead(sector: u64, count: usize, buffer: *u8) bool {
    if (count == 0) return true;
    const dev = blockDevice orelse return false;
    const d = dev.*;
    const qSize = d.queueSize;
    const descs = d.desc orelse return false;
    const avail = d.avail orelse return false;

    const hdr_ptr = kernel_alloc(@sizeOf(VirtioBlkOuthdr), 16) orelse return false;
    const status_ptr = kernel_alloc(1, 1) orelse return false;
    const hdr: *virtio.VirtioBlkOuthdr = @ptrCast(hdr_ptr);
    const status: *u8 = @ptrCast(status_ptr);

    hdr.* = VirtioBlkOuthdr{ .typ = VIRTIO_BLK_T_IN, .reserved = 0, .sector = sector };
    status.* = 0xFF;

    descs[0].addr = @as(u64, @intFromPtr(hdr));
    descs[0].len = @as(u32, @sizeOf(VirtioBlkOuthdr));
    descs[0].flags = virtio.VIRTQ_DESC_F_NEXT;
    descs[0].next = 1;

    descs[1].addr = @as(u64, @intFromPtr(buffer));
    descs[1].len = @as(u32, count * 512);
    descs[1].flags = virtio.VIRTQ_DESC_F_NEXT | virtio.VIRTQ_DESC_F_WRITE;
    descs[1].next = 2;

    descs[2].addr = @as(u64, @intFromPtr(status));
    descs[2].len = 1;
    descs[2].flags = virtio.VIRTQ_DESC_F_WRITE;
    descs[2].next = 0;

    // Update avail ring: ring starts 4 bytes after avail struct
    const ring_ptr: [*]u16 = @ptrCast(@as([*]u16, @ptrFromInt(@intFromPtr(avail) + 4)));
    const availIdx: usize = @intCast(avail.*.idx % @as(u16, qSize));
    ring_ptr[availIdx] = 0; // head of descriptor chain

    asm_volatile_barrier();

    avail.*.idx = avail.*.idx + 1;

    // Notify
    if (d.config.notify) |notify_ptr| {
        if (d.config.common) |common_ptr| {
            common_ptr.*.queue_select = 0;
            const off = @as(usize, common_ptr.*.queue_notify_off) * @as(usize, d.config.notify_off_multiplier);
            const notify_addr: *u16 = @ptrCast(@as(*u16, @ptrFromInt(@intFromPtr(notify_ptr) + off)));
            notify_addr.* = 0;
        }
    }

    var timeout: usize = 10_000_000;
    while (d.used.*.idx == d.lastUsedIdx and timeout > 0) {
        asm_pause();
        timeout -= 1;
    }

    if (timeout == 0) {
        return false;
    }

    // update lastUsedIdx
    d.lastUsedIdx = d.used.*.idx;

    return status.* == 0;
}
